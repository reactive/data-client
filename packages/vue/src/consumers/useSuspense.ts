import { ExpiryStatus } from '@data-client/core';
import type {
  EndpointInterface,
  Denormalize,
  Schema,
  FetchFunction,
  DenormalizeNullable,
  ResolveType,
} from '@data-client/core';
import {
  computed,
  watch,
  readonly,
  type DeepReadonly,
  type ComputedRef,
} from 'vue';

import { useController, injectState } from '../context.js';

export default function useSuspense<
  E extends EndpointInterface<
    FetchFunction,
    Schema | undefined,
    undefined | false
  >,
>(
  endpoint: E,
  ...args: readonly [...Parameters<E>]
): Promise<
  DeepReadonly<
    ComputedRef<
      E['schema'] extends undefined | null ? ResolveType<E>
      : Denormalize<E['schema']>
    >
  >
>;

export default function useSuspense<
  E extends EndpointInterface<
    FetchFunction,
    Schema | undefined,
    undefined | false
  >,
>(
  endpoint: E,
  ...args: readonly [...Parameters<E>] | readonly [null]
): Promise<
  DeepReadonly<
    ComputedRef<
      E['schema'] extends undefined | null ? ResolveType<E> | undefined
      : DenormalizeNullable<E['schema']>
    >
  >
>;

export default async function useSuspense(
  endpoint: any,
  ...args: any[]
): Promise<any> {
  const stateRef = injectState();
  const controller = useController();

  // Compute a key that changes when args change (including reactive props)
  const argsKey = computed(() =>
    args[0] !== null ? endpoint.key(...args) : '',
  );

  // Compute response meta reactively so we can respond to store updates
  const responseMeta = computed(() => {
    return argsKey.value ?
        controller.getResponseMeta(endpoint, ...args, stateRef.value)
      : null;
  });

  const maybeFetch = async () => {
    const currentKey = argsKey.value;
    if (!currentKey) return;
    const meta = responseMeta.value;
    if (!meta) return;
    const forceFetch = meta.expiryStatus === ExpiryStatus.Invalid;
    if (Date.now() <= meta.expiresAt && !forceFetch) return;
    await controller.fetch(endpoint, ...(args as any));
  };

  // Trigger on initial call
  await maybeFetch();

  // Watch for changes to key, expiry, or store state that require refetch
  watch(
    () => {
      const m = responseMeta.value;
      return m ?
          [m.expiresAt, m.expiryStatus, stateRef.value.lastReset, argsKey.value]
        : [argsKey.value];
    },
    () => {
      return maybeFetch();
    },
  );

  // Maintain GC refcounts on data mount/changes
  watch(
    () => responseMeta.value?.data,
    (_newVal, _oldVal, onCleanup) => {
      const decrement = responseMeta.value?.countRef();
      onCleanup(() => decrement?.());
    },
    { immediate: true },
  );

  // Return readonly computed ref - Vue automatically unwraps in templates and reactive contexts
  return readonly(computed(() => responseMeta.value?.data));
}
