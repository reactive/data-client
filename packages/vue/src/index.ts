// Vue Data Client - Coming Soon
// This package will provide Vue 3 composables for @data-client/core

export { Controller, ExpiryStatus, actionTypes } from '@data-client/core';
export type {
  EndpointExtraOptions,
  FetchFunction,
  ResolveType,
  EndpointInterface,
  EntityInterface,
  Queryable,
  SchemaArgs,
  Schema,
  SchemaClass,
  DenormalizeNullable,
  Denormalize,
  Normalize,
  NormalizeNullable,
  FetchAction,
  InvalidateAction,
  UnsubscribeAction,
  SubscribeAction,
  ResetAction,
  SetAction,
  SetResponseAction,
  ActionTypes,
  NetworkError,
  UnknownError,
  ErrorTypes,
  AbstractInstanceType,
  UpdateFunction,
  State,
  PK,
  Dispatch,
  Middleware,
  MiddlewareAPI,
  Manager,
  GCInterface,
  GCOptions,
  CreateCountRef,
  // used in Controller generic
  DataClientDispatch,
  GenericDispatch,
} from '@data-client/core';

// TODO: Add Vue-specific exports when implementation is ready
// export * from './composables/index.js';
// export * from './components/index.js';
// export * from './stores/index.js';
