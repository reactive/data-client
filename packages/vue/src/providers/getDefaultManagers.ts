import type { DevToolsConfig, Manager } from '@data-client/core';
import {
  NetworkManager,
  SubscriptionManager,
  PollingSubscription,
  DevToolsManager,
} from '@data-client/core';

/* istanbul ignore next */
/** Returns the default Managers used by DataProvider. */
let getDefaultManagers: (options?: GetManagersOptions) => Manager[] = (
  options = {},
) => {
  const { networkManager, subscriptionManager = PollingSubscription } = options;
  if (subscriptionManager === null) {
    return [constructManager(NetworkManager, networkManager ?? ({} as any))];
  }
  return [
    constructManager(NetworkManager, networkManager ?? ({} as any)),
    constructManager(SubscriptionManager, subscriptionManager),
  ];
};
/* istanbul ignore else */
if (process.env.NODE_ENV !== 'production') {
  getDefaultManagers = (options: GetManagersOptions = {}): Manager[] => {
    const {
      devToolsManager,
      networkManager,
      subscriptionManager = PollingSubscription,
    } = options;
    if (networkManager === null) {
      console.error('Disabling NetworkManager is not allowed.');
      // fall back to default options
    }
    const nm = constructManager(NetworkManager, networkManager ?? ({} as any));
    const managers: Manager[] = [nm];
    if (subscriptionManager !== null) {
      managers.push(constructManager(SubscriptionManager, subscriptionManager));
    }
    if (devToolsManager !== null) {
      let dtm: DevToolsManager;
      if (devToolsManager instanceof DevToolsManager) {
        dtm = devToolsManager;
      } else {
        dtm = new DevToolsManager(
          devToolsManager as any,
          nm.skipLogging.bind(nm),
        );
      }
      managers.unshift(dtm);
    }
    return managers;
  };
}
export { getDefaultManagers };

function constructManager<M extends { new (...args: any): Manager }>(
  Mgr: M,
  optionOrInstance: InstanceType<M> | ConstructorArgs<M>,
): InstanceType<M> {
  if (optionOrInstance instanceof Mgr) {
    return optionOrInstance as InstanceType<M>;
  }
  return new Mgr(optionOrInstance as any) as InstanceType<M>;
}

export type GetManagersOptions = {
  devToolsManager?: DevToolsManager | DevToolsConfig | null;
  networkManager?:
    | NetworkManager
    | ConstructorArgs<typeof NetworkManager>
    | null;
  subscriptionManager?:
    | SubscriptionManager
    | ConstructorArgs<typeof SubscriptionManager>
    | null;
};

export type ConstructorArgs<T extends { new (...args: any): any }> =
  T extends new (options: infer O) => any ? O : never;
