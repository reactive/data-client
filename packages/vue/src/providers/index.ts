export { getDefaultManagers } from './getDefaultManagers.js';
export { createDataClient } from './createDataClient.js';
export { DataClientPlugin } from './DataClientPlugin.js';
