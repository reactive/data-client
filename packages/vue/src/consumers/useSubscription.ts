import type {
  EndpointInterface,
  Schema,
  FetchFunction,
} from '@data-client/core';
import { computed, unref, watch } from 'vue';

import { useController } from '../context.js';
import type { MaybeRefsOrGettersNullable } from '../types.js';

/**
 * Keeps a resource fresh by subscribing to updates.
 * Mirrors React hook API. Pass `null` as first arg to unsubscribe.
 * @see https://dataclient.io/docs/api/useSubscription
 */
export default function useSubscription<
  E extends EndpointInterface<
    FetchFunction,
    Schema | undefined,
    undefined | false
  >,
>(
  endpoint: E,
  ...args: MaybeRefsOrGettersNullable<Parameters<E>> | readonly [null]
) {
  const controller = useController();

  // Track top-level reactive args (Refs are unwrapped). This allows props/refs to trigger resubscribe.
  const resolvedArgs = computed(() => args.map(a => unref(a as any)) as any);
  const key = computed(() => {
    if (resolvedArgs.value[0] === null) return '';
    return endpoint.key(...(resolvedArgs.value as readonly [...Parameters<E>]));
  });

  // Subscribe when key exists; unsubscribe on change or unmount
  watch(
    key,
    (_newKey, _oldKey, onCleanup) => {
      if (!key.value) return;
      const cleanedArgs = resolvedArgs.value as readonly [...Parameters<E>];
      controller.subscribe(endpoint, ...cleanedArgs);
      onCleanup(() => {
        controller.unsubscribe(endpoint, ...cleanedArgs);
      });
    },
    { immediate: true },
  );
}
