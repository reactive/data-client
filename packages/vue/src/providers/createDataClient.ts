import {
  initialState as defaultState,
  Controller as DataController,
  applyManager,
  initManager,
  createReducer,
} from '@data-client/core';
import type { State, Manager, GCInterface } from '@data-client/core';
import { provide, shallowRef, type ShallowRef, type App } from 'vue';

import { ControllerKey, StateKey } from '../context.js';
import { getDefaultManagers } from './getDefaultManagers.js';

export interface ProvideOptions {
  managers?: Manager[];
  initialState?: State<unknown>;
  Controller?: typeof DataController;
  gcPolicy?: GCInterface;
  app?: App;
}

export interface ProvidedDataClient {
  controller: InstanceType<typeof DataController>;
  /** Optimistic overlay state ref provided to consumers */
  stateRef: ShallowRef<State<unknown>>;
  /** Start the provider (called on mount) */
  start: () => void;
  /** Stop the provider (called on unmount) */
  stop: () => void;
}

/**
 * Core provider logic that can be used by both composable and plugin.
 * This function handles the actual setup of the data client without Vue-specific concerns.
 */
export function createDataClient(
  options: ProvideOptions = {},
): ProvidedDataClient {
  const { Controller = DataController, gcPolicy, app } = options;

  // stable singletons for this provider scope
  const controller = new Controller({ gcPolicy });
  const managers = options.managers ?? getDefaultManagers();
  const baseInitial = options.initialState ?? (defaultState as State<unknown>);

  // init managers (run on mount/unmount)
  const mgrEffect = initManager(managers, controller, baseInitial);

  // build middlewares and bind to controller
  const middlewares = applyManager(managers, controller);

  // reducer and state management
  const reducer = createReducer(controller);

  // base state (no optimistic overlay) exposed to controller.getState via middleware API
  const baseStateRef = shallowRef<State<unknown>>(baseInitial);
  // optimistic/effective state provided to consumers
  const stateRef = shallowRef<State<unknown>>(computeOptimistic(baseInitial));

  type Dispatch = (action: any) => Promise<void>;

  let resolveNext: (() => void) | null = null;
  const waitForCommit = () =>
    new Promise<void>(resolve => {
      resolveNext = resolve;
    });

  const realDispatch: Dispatch = async action => {
    const nextBase = reducer(baseStateRef.value, action);
    baseStateRef.value = nextBase;
    stateRef.value = computeOptimistic(nextBase);
    if (resolveNext) {
      const r = resolveNext;
      resolveNext = null;
      r();
    }
  };

  const getState = () => baseStateRef.value;

  // compose middlewares similar to redux
  const chain = middlewares.map(mw =>
    mw({
      getState,
      dispatch: (a: any) => outerDispatch(a),
    } as any),
  );
  const compose = (fns: ((arg: any) => any)[]) => (initial: any) =>
    fns.reduceRight((v, f) => f(v), initial);
  const outerDispatch = compose(chain)(async (action: any) => {
    const promise = waitForCommit();
    await realDispatch(action);
    return promise;
  });

  // wire controller
  controller.bindMiddleware({
    getState,
    dispatch: (a: any) => outerDispatch(a),
  } as any);

  // setup lifecycle management
  let cleanup: void | (() => void);
  const start = () => {
    cleanup = mgrEffect();
  };
  const stop = () => {
    if (cleanup) cleanup();
  };

  // provide to children using Vue's provide or app.provide
  if (app) {
    // Plugin mode: use app.provide
    app.provide(StateKey, stateRef);
    app.provide(ControllerKey, controller as any);
  } else {
    // Composable mode: use provide (must be called within setup)
    provide(StateKey, stateRef);
    provide(ControllerKey, controller as any);
  }

  return {
    controller: controller as any,
    stateRef,
    start,
    stop,
  };

  function computeOptimistic(state: State<unknown>): State<unknown> {
    // mirror React's optimistic overlay
    // reduce over pending optimistic actions to derive the effective state
    // reducer is stable from closure
    return state.optimistic.reduce(reducer as any, state);
  }
}
