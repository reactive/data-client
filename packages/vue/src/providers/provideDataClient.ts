import {
  initialState as defaultState,
  Controller as DataController,
  applyManager,
  initManager,
  createReducer,
} from '@data-client/core';
import type { State, Manager, GCInterface } from '@data-client/core';
import {
  onMounted,
  onUnmounted,
  provide,
  shallowRef,
  type ShallowRef,
} from 'vue';

import { ControllerKey, StateKey } from '../context.js';
import { getDefaultManagers } from './getDefaultManagers.js';

export interface ProvideOptions {
  managers?: Manager[];
  initialState?: State<unknown>;
  Controller?: typeof DataController;
  gcPolicy?: GCInterface;
}

export interface ProvidedDataClient {
  controller: InstanceType<typeof DataController>;
  /** Optimistic overlay state ref provided to consumers */
  stateRef: ShallowRef<State<unknown>>;
}

/**
 * Provide/inject setup for @data-client/vue. Call inside setup() of your root component.
 * Mirrors React DataProvider but as a composable.
 */
export function provideDataClient(
  options: ProvideOptions = {},
): ProvidedDataClient {
  const { Controller = DataController, gcPolicy } = options;

  // stable singletons for this provider scope
  const controller = new Controller({ gcPolicy });
  const managers = options.managers ?? getDefaultManagers();
  const baseInitial = options.initialState ?? (defaultState as State<unknown>);

  // init managers (run on mount/unmount)
  const mgrEffect = initManager(managers, controller, baseInitial);

  // build middlewares and bind to controller
  const middlewares = applyManager(managers, controller);

  // reducer and state management
  const reducer = createReducer(controller);

  // base state (no optimistic overlay) exposed to controller.getState via middleware API
  const baseStateRef = shallowRef<State<unknown>>(baseInitial);
  // optimistic/effective state provided to consumers
  const stateRef = shallowRef<State<unknown>>(computeOptimistic(baseInitial));

  type Dispatch = (action: any) => Promise<void>;

  let resolveNext: (() => void) | null = null;
  const waitForCommit = () =>
    new Promise<void>(resolve => {
      resolveNext = resolve;
    });

  const realDispatch: Dispatch = async action => {
    const nextBase = reducer(baseStateRef.value, action);
    baseStateRef.value = nextBase;
    stateRef.value = computeOptimistic(nextBase);
    if (resolveNext) {
      const r = resolveNext;
      resolveNext = null;
      r();
    }
  };

  const getState = () => baseStateRef.value;

  // compose middlewares similar to redux
  const chain = middlewares.map(mw =>
    mw({
      getState,
      dispatch: (a: any) => outerDispatch(a),
    } as any),
  );
  const compose = (fns: ((arg: any) => any)[]) => (initial: any) =>
    fns.reduceRight((v, f) => f(v), initial);
  const outerDispatch = compose(chain)(async (action: any) => {
    const promise = waitForCommit();
    await realDispatch(action);
    return promise;
  });

  // wire controller
  controller.bindMiddleware({
    getState,
    dispatch: (a: any) => outerDispatch(a),
  } as any);

  // provide to children
  provide(StateKey, stateRef);
  provide(ControllerKey, controller as any);

  // run managers after mount and cleanup on unmount
  let cleanup: void | (() => void);
  onMounted(() => {
    cleanup = mgrEffect();
  });
  onUnmounted(() => {
    if (cleanup) cleanup();
  });

  return { controller: controller as any, stateRef };

  function computeOptimistic(state: State<unknown>): State<unknown> {
    // mirror Reactâ€™s optimistic overlay
    // reduce over pending optimistic actions to derive the effective state
    // reducer is stable from closure
    return state.optimistic.reduce(reducer as any, state);
  }
}

export default provideDataClient;
