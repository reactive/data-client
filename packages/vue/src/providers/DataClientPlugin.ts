import { Controller } from '@data-client/core';
import type { App } from 'vue';

import { createDataClient, type ProvideOptions } from './createDataClient.js';

/**
 * Vue 3 Plugin for Reactive Data Client
 *
 * Usage:
 * ```ts
 * import { createApp } from 'vue';
 * import { DataClientPlugin } from '@data-client/vue';
 *
 * const app = createApp(App);
 * app.use(DataClientPlugin, {
 *   managers: getDefaultManagers(),
 *   initialState: customInitialState,
 * });
 * app.mount('#app');
 * ```
 */
export const DataClientPlugin = {
  install(app: App, options: ProvideOptions = {}) {
    const provider = createDataClient({ ...options, app });

    app.config.globalProperties.$dataClient = provider.controller;

    /* istanbul ignore if */
    if (process.env.NODE_ENV === 'development') {
      console.info('Starting DataClientPlugin');
    }

    provider.start();

    app.onUnmount(() => {
      /* istanbul ignore if */
      if (process.env.NODE_ENV === 'development') {
        console.info('Stopping DataClientPlugin');
      }

      provider.stop();
    });

    return provider;
  },
};

// install global property types

declare module 'vue' {
  interface ComponentCustomProperties {
    $dataClient: Controller;
  }
}
