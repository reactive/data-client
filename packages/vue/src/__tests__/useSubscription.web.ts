import { mount } from '@vue/test-utils';
import nock from 'nock';
import { defineComponent, h, nextTick, Suspense } from 'vue';

// Endpoints/entities from React subscriptions test
import {
  PollingArticleResource,
  ArticleResource,
} from '../../../../__tests__/new';
import useSubscription from '../consumers/useSubscription';
import useSuspense from '../consumers/useSuspense';
import { provideDataClient } from '../providers/provideDataClient';

describe('vue useSubscription()', () => {
  const payload = {
    id: 5,
    title: 'hi ho',
    content: 'whatever',
    tags: ['a', 'best', 'react'],
  };

  // Mutable payload to simulate server-side updates with polling
  let currentPollingPayload: typeof payload = { ...payload };

  async function flushUntil(
    wrapper: any,
    predicate: () => boolean,
    tries = 100,
  ) {
    for (let i = 0; i < tries; i++) {
      if (predicate()) return;
      await Promise.resolve();
      await nextTick();
      await new Promise(resolve => setTimeout(resolve, 0));
    }
  }

  beforeAll(() => {
    // Global network stubs reused by tests
    nock(/.*/)
      .persist()
      .defaultReplyHeaders({
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Headers': 'Access-Token',
        'Content-Type': 'application/json',
      })
      .options(/.*/)
      .reply(200)
      // ArticleResource and PollingArticleResource both hit /article/:id
      .get(`/article/${payload.id}`)
      .reply(200, () => currentPollingPayload)
      .put(`/article/${payload.id}`)
      .reply(200, (uri, requestBody: any) => ({
        ...currentPollingPayload,
        ...requestBody,
      }));
  });

  afterAll(() => {
    nock.cleanAll();
  });

  const ProvideWrapper = defineComponent({
    name: 'ProvideWrapper',
    setup(_props, { slots, expose }) {
      const { controller } = provideDataClient();
      expose({ controller });
      return () =>
        h(
          Suspense,
          {},
          {
            default: () => (slots.default ? slots.default() : null),
            fallback: () => h('div', { class: 'fallback' }, 'Loading'),
          },
        );
    },
  });

  const ArticleComp = defineComponent({
    name: 'ArticleComp',
    props: { active: { type: Boolean, default: true } },
    async setup(props) {
      // Subscribe BEFORE any await to preserve current instance for inject()
      useSubscription(
        PollingArticleResource.get,
        props.active ? { id: payload.id } : (null as any),
      );
      const article = await useSuspense(PollingArticleResource.get, {
        id: payload.id,
      });
      return () =>
        h('div', [
          h('h3', (article as any).value.title),
          h('p', (article as any).value.content),
        ]);
    },
  });

  it('subscribes and re-renders on updates (simulated poll)', async () => {
    currentPollingPayload = { ...payload };

    const wrapper = mount(ProvideWrapper, {
      slots: { default: () => h(ArticleComp, { active: true }) },
    });

    // Initially should render fallback while Suspense is pending
    expect(wrapper.find('.fallback').exists()).toBe(true);

    // Flush initial fetch
    await flushUntil(wrapper, () => wrapper.find('h3').exists());

    // Verify initial values
    expect(wrapper.find('h3').text()).toBe(payload.title);
    expect(wrapper.find('p').text()).toBe(payload.content);

    // Simulate a polling update by changing server payload and manually fetching
    const updatedTitle = payload.title + ' fiver';
    currentPollingPayload = { ...payload, title: updatedTitle } as any;
    const exposed: any = wrapper.vm as any;
    await exposed.controller.fetch(PollingArticleResource.get, {
      id: payload.id,
    });

    await flushUntil(wrapper, () => wrapper.find('h3').text() === updatedTitle);
    expect(wrapper.find('h3').text()).toBe(updatedTitle);
  });

  it('can subscribe to endpoint without pollFrequency (no-op) and render', async () => {
    // Minimal component that subscribes to non-polling endpoint
    const NoFreqComp = defineComponent({
      name: 'NoFreqComp',
      async setup() {
        // Subscribe first (no poller attached)
        useSubscription(ArticleResource.get, { id: payload.id } as any);
        // Then resolve suspense for stable render
        const article = await useSuspense(ArticleResource.get, {
          id: payload.id,
        });
        return () => h('div', (article as any).value.title);
      },
    });

    const wrapper = mount(ProvideWrapper, {
      slots: { default: () => h(NoFreqComp) },
    });

    await flushUntil(wrapper, () => wrapper.text() !== '');
    expect(wrapper.text()).not.toEqual('');
  });
});
