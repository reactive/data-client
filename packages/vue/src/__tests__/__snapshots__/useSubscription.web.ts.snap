// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`vue useSubscription() can subscribe to endpoint without pollFrequency (no-op) and render 1`] = `
[
  [Error: frequency needed for polling subscription],
]
`;
